packages/@lwc/engine-dom/src/custom-elements/create-custom-element.ts

```diff
@@ -29,46 +29,55 @@ const nativeLifecycleElementsToUpgradedByLWC = new WeakMap<HTMLElement, boolean>
 
 let elementBeingUpgradedByLWC = false;
 
+// TODO [#2972]: this class should expose observedAttributes as necessary
+class BaseUpgradableConstructor extends HTMLElement {
+    constructor(upgradeCallback: LifecycleCallback, useNativeLifecycle: boolean) {
+        super();
+
+        if (useNativeLifecycle) {
+            // When in native lifecycle mode, we need to keep track of instances that were created outside LWC
+            // (i.e. not created by `lwc.createElement()`). If the element uses synthetic lifecycle, then we don't
+            // need to track this.
+            nativeLifecycleElementsToUpgradedByLWC.set(this, elementBeingUpgradedByLWC);
+        }
+
+        // If the element is not created using lwc.createElement(), e.g. `document.createElement('x-foo')`,
+        // then elementBeingUpgradedByLWC will be false
+        if (elementBeingUpgradedByLWC) {
+            upgradeCallback(this);
+        }
+        // TODO [#2970]: LWC elements cannot be upgraded via new Ctor()
+        // Do we want to support this? Throw an error? Currently for backwards compat it's a no-op.
+    }
+}
+
+for (const [propName, callback] of entries(LIFECYCLE_CALLBACKS)) {
+    // @ts-expect-error type-mismatch
+    BaseUpgradableConstructor.prototype[propName] = function () {
+        // If the element is in the WeakMap (i.e. it's marked as native lifecycle), and if it was upgraded by LWC,
+        // then it can use native lifecycle
+        if (isTrue(nativeLifecycleElementsToUpgradedByLWC.get(this))) {
+            callback(this);
+        }
+    };
+}
```

Comment:
This class is basically hoisted unmodified from the old `UpgradableConstructor`. The only difference is that it's missing the `formAssociated`, which has to be different per-class.